### ISSUE TITLE
In-project custom blocks/functions created in JS shown in blocks

### ISSUE THREAD

**Issue by jaustin** [Microsoft/pxt#882]
It would be really nice if functions created locally (IE in the script file) in Javascript could be displayed as 'code' or 'custom' blocks.

This would allow users who have transitioned to JS a path back to blocks without requiring a way of expressing arbitrary JS inside blocks. Also advanced users could make cool things that blocks users can consume

Here's a quick mockup just to get the idea...

In this example, we imagine that there's a function that sends a number by flashing an LED in a specific way, LEDSendNumber, and that we want to call that when ButtonA is pressed.
![image](https://cloud.githubusercontent.com/assets/147113/20981223/a3af8dce-bcab-11e6-91c9-1abb43b82712.png)
There a few things going on here:
 * New item in the list of block categories "In this script" (there could even be a 'new block' option in here too ;) )
 * The block inside the button A handler is a different colour to others and contains the fact that it's local somehow (local. used here, but not at all attached to that), but is drawn as accepting an argument
 * The help for the block has a description and tells the user it's local/gives a chance to edit it

The Javascript might look like (super naive implementation just to give the point):
```javascript
input.onButtonPressed(Button.A, () => {
    LEDSendNumber(20)
})
function LEDSendNumber(num: number) {
    basic.showNumber(num)
    //Send start sequence
    pins.digitalWritePin(DigitalPin.P1, 1)
    basic.pause(100)
    pins.digitalWritePin(DigitalPin.P1, 0)
    //Flash the right number of times
    for (let index = 0; index <= num; index++) {
        basic.pause(10)
        pins.digitalWritePin(DigitalPin.P1, 1)
        basic.pause(10)       
        pins.digitalWritePin(DigitalPin.P1, 0)
    }
    //Send end sequence
    pins.digitalWritePin(DigitalPin.P1, 1)
    basic.pause(200)
    pins.digitalWritePin(DigitalPin.P1, 0)
}
```
Actually because this function is only expressed in terms of blocks, it would be really nice to be able to edit it in block-land too, but I know that's potentially a big step compared to just showing the block:
![image](https://cloud.githubusercontent.com/assets/147113/20982076/f92c7b24-bcae-11e6-884a-0f36816ee4d2.png)

**Comment by whaleygeek** [comment]
So basically what you are suggesting, is to add function definition (and user function activation) to the visual blocks view? A bit like the 'new block' block in fig3 in this paper: http://www.sciencedirect.com/science/article/pii/S1877050915025879

**Comment by d-f-booth** [comment]
Looking at this ticket 882 and 325 - I think that it would be a significant advance if block scripts could contain the ability to write subroutines with the ability to have dummy arguments and the option to pass on a modified variable. I have found several instances where I have had to copy/repeat sections of block code just because one or two variable in the section have changed but the function performed on them is the same.

In my opinion the ability to write subroutines is one of the major features of programming allowing unnecessary work to be avoided and ultimately leading to the concept of Classes.  So from an educational point of view it is a very powerful concept that should be introduced as soon as a student has grasped the basics of writing a program to control an activity.
David Booth
